### PR TITLE
Draft: nix deployment

### PR DESCRIPTION
**Pull request by happysalada**
I've got an initial deploment half working at https://bonfire.union.rocks/

I started this PR as a draft so we can discuss the missing steps.
(I can clean the commits later)
- The migrate task in production is missing. The readme talked about `Bonfire.Repo.ReleaseTasks.migrate`, but it doesn't seem to exist anymore (maybe I missed it). I can propose a propose a production migrate task, but perhaps the approach you want to take here is different? I'm saying that since you were mentioning not knowing the schema at runtime even.
- I've had to remove the RELEASING variable. Systemd can only pass string environment variables, and so it was getting parsed as the string "1" which replaced the runtime.exs env vars.
- I'm including configurable out of the box the PORT, POSTGRES_DB, HOSTNAME, and adding an environment file for secrets 
SECRET_KEY_BASE, SIGNING_SALT, ENCRYPTION_SALT, RELEASE_COOKIE. I'm just mentioning this in case you feel there is the need for additional configuration.
- I tried adding the valueflows dependency, mix complained of missing bonfire_quantity, so I added it, then it complained of missing graphql dependency so I added on production. This is perhaps not what you intended, so I am mentioning it here.
- I tried adding valueflows_ui, but compilation failed from a missing Forms module, and I didn't dig further.

This is a little "rough", but almost working. The main blocker being the db migrate task.

### CODE REVIEW

**Comment by mayel** [comment]
I've added a few review comments. Regarding the migrate task, can you share details about your error, because it should be there:
https://www.github.com/bonfire-networks/bonfire_common/tree/main/lib%2Frepo%2Frelease_tasks.ex

**Comment by happysalada** [comment]
Thanks for the feedback. I think you need to click on review for the comments to appear.

I've checked on production, and I misdiagnosed the problem. The migrations ran properly, but I'm still getting a missing table error.
```
query: DELETE FROM \"public\".\"oban_jobs\" AS o0 USING (SELECT so0.\"id\" AS \"id\" FROM \"public\".\"oban_jobs\" AS so0 WHERE (so0.\"state\" IN ('completed','discarded')) AND (so0.\"attempted_at\" < $1) LIMIT $2) AS s1 WHERE (o0.\"id\" = s1.\"id\")"}

* (Postgrex.Error) ERROR 42P01 (undefined_table) relation \"public.oban_jobs\" does not exist"}
```
do I need a particular dependency for the oban_jobs table to be created?

**Comment by mayel** [comment]
Oban migrations should be included by default, so that may be a sign migrations didn't run correctly...

File: config/config.exs [hunk 1]
```diff
@@ -15,9 +15,9 @@ import_config "bonfire_classify.exs"
 
 # import_config "bonfire_publisher_thesis.exs"
 import_config "bonfire_fail.exs"
-# import_config "bonfire_quantify.exs"
+import_config "bonfire_quantify.exs"
```
[the remark this answers is not in the document]
Comment (happysalada) [reply config/config.exs:18]: I see, I didn't know there was a special branch for it. I'll give it a go, thanks!

**Comment by happysalada** [comment]
Ok, I'll try to clear the database, and run the migrations separately.
I'll use the valueflows branch now that I know it exist.
I'll update with the results.

File: mix.exs [hunk 2]
```diff
@@ -66,17 +64,21 @@ defmodule Bonfire.MixProject do
       {:phoenix_live_reload, "~> 1.2", only: :dev},
       {:exsync, "~> 0.2", only: :dev},
 
+      # needed for bonfire quantify
+      {:bonfire_api_graphql,
```
[the remark this answers is not in the document]
Comment (happysalada) [reply mix.exs:68]: If I remember correctly, it was when I included quantify that mix started complaining. I included it mechanically to just move on to the next step.
I'll use the prefered branch, so I can get rid of the dependency noise.
Thank you for the explanation.

File: config/runtime.exs [hunk 3]
```diff
@@ -5,17 +5,16 @@
 import Config
 
 if config_env() == :prod do
-
   host = System.get_env("HOSTNAME", "localhost")
   port = String.to_integer(System.get_env("PORT", "4000"))
 
-  System.get_env("RELEASING") || System.get_env("DATABASE_URL") || (System.get_env("POSTGRES_DB") && System.get_env("POSTGRES_PASSWORD")) ||
```
[the remark this answers is not in the document]
Comment (happysalada) [reply config/runtime.exs:12]: So the problem I encountered is that I can only supply strings to systemd en variables.
That in turns gets parsed into "1" which replaces thee value themselves.

Now that I think about it, I should actually not set RELEASING. I understood that the other way around. I'll do another try without setting RELEASING.
Thanks for the clarification.

File: nix/package.nix [hunk 4]
```diff
@@ -0,0 +1,63 @@
+{ pkgs, self }:
+let
+  beamPackages = with pkgs; beam.packagesWith beam.interpreters.erlang;
+in
+beamPackages.mixRelease rec {
+  pname = "bonfire";
+  version = "1.0.0";
```
[the remark this answers is not in the document]
Comment (happysalada) [reply nix/package.nix:7]: Yeah, you're right, ideally it should be set automatically. I'll try to look for a way.

File: config/runtime.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayel) [reply config/runtime.exs:12]: 👍

File: config/config.exs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mayel) [reply config/config.exs:18]: 👍 I'll copy the nix config back into main once we've got it merged there.

**Comment by happysalada** [comment]
One thing I forgot to add.
Once this is a little more stable, distributing it into nixpkgs can enable non NixOS linux to get the binary. After that it's just a matter of including an example systemd service in the repo so that non NixOS linux can deploy the release easily. nix would just be used as a convenient way to get the right binary.
This might be more popular than installing nixos and running on it.

**Comment by jjl** [comment]
> One thing I forgot to add.
> Once this is a little more stable, distributing it into nixpkgs can enable non NixOS linux to get the binary. After that it's just a matter of including an example systemd service in the repo so that non NixOS linux can deploy the release easily. nix would just be used as a convenient way to get the right binary.
> This might be more popular than installing nixos and running on it.

That sounds convenient, but given how reluctant erlang is to run on systems that differ from where it was compiled, i'm a little sceptical.

**Comment by happysalada** [comment]
> > One thing I forgot to add.
> > Once this is a little more stable, distributing it into nixpkgs can enable non NixOS linux to get the binary. After that it's just a matter of including an example systemd service in the repo so that non NixOS linux can deploy the release easily. nix would just be used as a convenient way to get the right binary.
> > This might be more popular than installing nixos and running on it.
> 
> That sounds convenient, but given how reluctant erlang is to run on systems that differ from where it was compiled, i'm a little sceptical.

I might be over optimistic here 😄 
I haven't seen it in the wild, just something that came to me. I'll check up more on it.

**Comment by mayel** [comment]
@happysalada you may want to hold on a sec, as we're improving the config setup by introducing BONFIRE_FLAVOUR and the secondary branch is going away.

**Comment by happysalada** [comment]
sure, no worries, just ping me when you are ready.

**Comment by mayel** [comment]
@happysalada it's still a work in progress, but if you want to look at it again, I've added some docs and improved the makefile

**Comment by happysalada** [comment]
Thanks, I'll come back to you as soon as I've had time to time to test it.

**Pull request closed without merging**